Add Route53

kagux opened the pull request:
Hi,

Thanks a lot for your work on this project! 
I'd like to contribute with Route53 endpoints. This PR is just a single endpoint to list hosted zone and  I will push a few more later today or tomorrow. But I'd love to get your feedback on the direction I took with this implementation.

Endpoints:
- list_hosted_zones
- create_hosted_zone
- delete_hosted_zone
- change_record_sets
- list_record_sets

benwilson512 commented:
Hey man lots of good stuff here, thanks for the PR! I'll try to get a review in here later today.

kagux commented:
I've added a couple of new endpoints

kagux commented on lib/ex_aws/utils.ex:
```diff
@@ -100,4 +100,6 @@ defmodule ExAws.Utils do
     greg_secs - @seconds_0_to_1970
   end
+
+  def uuid, do: DateTime.utc_now |> :erlang.phash2
```
I think it is good enough for a simple uuid. If we were to consider improving it, then using external package like https://hex.pm/packages/uuid could be a better alternative. I didn't want to bring in a package unless really needed.

benwilson512 replied:
I'm gonna switch this to `:erlang.unique_integer([:positive])` since concurrent requests could easily call this function at the same time.

benwilson512 replied:
Actually never mind, those can repeat over instances. I suppose concurrently creating hosted zones is unlikely so we'll stick with the current implementation.

kagux commented:
Added error response handling

benwilson512 commented on lib/ex_aws/parsers.ex:
```diff
@@ -0,0 +1,18 @@
+                          type: ~x"./Error/Type/text()"s,
+                          code: ~x"./Error/Code/text()"s,
+                          message: ~x"./Error/Message/text()"s,
+                          detail: ~x"./Error/Detail/text()"s)
```
Is this correct? Every service that uses XML has this pattern? Including S3?

I'm a little reluctant to do this because AWS services are not known for being particularly consistent with respect to one another. There is some level of consistency that exists with respect to a given Operation type like Query or JSON or RestQuery but across those there is very little.

kagux replied:
This is correct for `Route53`, `SNS` and `SQS`. Error response formats are identical. Relevant links:

- http://docs.aws.amazon.com/Route53/latest/APIReference/requests-rest-responses.html
-http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/UnderstandingResponses.html

I didn't find reference for SNS, but it is exactly the same and my initial trigger to extract this were identical test examples. 
How would you suggest renaming this module?

benwilson512 replied:
Fair enough. I think it should likely go under `ExAws.Operation.Query.Parser` because the parsing it does seems specific to that particular operation format. It does not relate to either S3 or say Dynamo.

benwilson512 commented on lib/ex_aws/operation/rest_query.ex:
```diff
@@ -3,37 +3,24 @@ defmodule ExAws.Operation.RestQuery do
-    ]
-
-    ExAws.Request.request(operation.http_method, url(config, operation.path), query, headers, config, operation.service)
+    headers = []
```
Why are we not setting the header here now?

kagux replied:
it became more complicated as headers vary depending on request method and body type. It works works for all current use cases without headers. So I decided to skip it to avoid bloating PR and make a separate PR to set correct headers.

benwilson512 commented on lib/ex_aws/utils.ex:
```diff
@@ -100,4 +100,15 @@ defmodule ExAws.Utils do
+       true -> {k, v}
+     end
+    end)
+  end
```
I think this would be better as:
```
def rename_keys(params, mapping) do
  mapping = Map.new(mapping)
  Enum.map(params, fn {k, v} ->
    {Map.get(mapping, k, k), v}
  end)
end
```

kagux replied:
nice one, thanks, I'll update!

benwilson512 commented on lib/ex_aws/request/url.ex:
```diff
@@ -0,0 +1,23 @@
+    url |> Map.put(:scheme, scheme)
+  end
+  defp normalize_scheme(url), do: url
+end
```
I appreciate moving this into its own module, I think there's a lot of potential here for cleaning up code from other operations. However if we're gonna have a module named `ExAws.Request.URL` we should make sure that it works for all services. If it does not, it belongs under the operation namespace not the request namespace.

Could you try and see if you can use this with all operations?

kagux replied:
Sure, I'll replace URL manipulation with this module for all operations.

kagux commented:
I've addressed all your comments except for `headers` list. I think it shouldn't be done within the scope of current PR. 
To set correct headers we either have to pass content type to operation or set it based on combination of request method and body content.

benwilson512 commented:
There's one other relatively minor issue I'm noticing: Route 53 isn't actually a Rest Query operation, it's a Rest XML operation. This explains why some of the header stuff wasn't making sense. While we don't actually have a Rest XML operation yet, it should be pretty easy to build by just cloning Rest Query version. This lets us set the headers on the rest query back to what they were.

kagux commented:
it depends on how you define each operation. To be honest, I don't think that splitting would solve the problem. In original rest query implementation, header was always set to "application/x-www-form-urlencoded", which technically wasn't correct.  Adding XML POST body exposed this issue.  
Splitting out XML would introduce quite some duplication and would solve headers issue only for one particular case. 

 If you insist, I'll clone a new operation. I do think that splitting header logic from operation or making content type part of operation would be a  better long term solution. 

How do you want me to proceed?

kagux commented:
Updated `ExAws.Operation.Query.Parser` module. Headers should be the only thing left pending in this PR.

kagux commented:
@benwilson512, how do you want to proceed?

kagux commented:
I've added new endpoint `get_record_sets` and rebased to remove merge conflicts on mix.lock

kagux commented:
My last change is related to this issue in `sweet_xml` https://github.com/awetzel/sweet_xml/pull/20. Text nodes with `&quot;` are split in `xmerl_scan`. They fixed it in `sweet_xml` for a single node. In our case, we want a list of strings and that ends up splitting some nodes into multiple.
Amazon wraps TXT zone records in quotes and as a result triggers this issue. I couldn't figure any other way around this issue but decoding quotes.

kagux commented:
hey @benwilson512, 
I was wondering if we could wrap up with this PR. We've been using PR-fork in production for last month with no issues.
Please, let me know if there's still a blocking issue in current PR state.
